cmd/faucet-v2/main.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/types/bech32"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/sethvargo/go-envconfig"
+)
+
+type Config struct {
+	CliName        string `env:"CLI_NAME, default=wardend"`
+	ChainID        string `env:"CHAIN_ID, default=warden"`
+	KeyringBackend string `env:"KEYRING_BACKEND, default=test"`
+	Node           string `env:"NODE, default=http://localhost:26657"`
+	SendDenom      string `env:"DENOM, default=10000000uward"`
+	AccountName    string `env:"ACCOUNT_NAME, default=shulgin"`
+	Mnemonic       string `env:"MNEMONIC"`
+	HDPath         string `env:"HD_PATH, default=m/44'/118'/0'/0/0"`
+	Fees           string `env:"FEES, default=1uward"`
+	OtherFlags     string `env:"OTHER_FLAGS"`
+
+	Cooldown      time.Duration `env:"COOLDOWN, default=12h"`
+	BatchInterval time.Duration `env:"BATCH_INTERVAL, default=6s"`
+	BatchLimit    int           `env:"BATCH_LIMIT, default=10"`
+
+	RecaptchaSiteKey      string  `env:"RECAPTCHA_SITE_KEY"`
+	RecaptchaSecret       string  `env:"RECAPTCHA_SECRET_KEY"`
+	RecaptchaMinimumScore float64 `env:"RECAPTCHA_MINIMUM_SCORE, default=0.8"`
+}
+
+type Env struct {
+	Logger         *slog.Logger
+	AddressLimiter *Limiter
+	IPLimiter      *Limiter
+	Client         *Client
+}
+
+func main() {
+	ctx := context.Background()
+	var c Config
+	if err := envconfig.Process(ctx, &c); err != nil {
+		log.Fatal(err)
+	}
+
+	logger := slog.New(slog.NewJSONHandler(os.Stderr, nil))
+
+	client, err := NewClient(ctx, logger.With("module", "client"), c)
+	if err != nil {
+		log.Fatalf("can't init client: %v", err)
+	}
+
+	env := Env{
+		Logger:         logger,
+		AddressLimiter: NewLimiter(c.Cooldown),
+		IPLimiter:      NewLimiter(c.Cooldown),
+		Client:         client,
+	}
+
+	http.HandleFunc("/", formHandler(c))
+	http.HandleFunc("/submit", submitHandler(c, env))
+	http.Handle("/metrics", promhttp.Handler())
+	logger.Info("faucet listening at :8000")
+	if err := http.ListenAndServe(":8000", nil); err != nil {
```

Comment:
I think it would be good to have a timeout here?

Reply:
sorry, where?

Reply:
So that the clients cannot uphold a connection indefenetively

Reply:
I added `ReadHeaderTimeout`, is this enough or should I set more timeouts?

Reply:
lgtm